plugin/pkg/managedseed/validator/admission.go

```diff
@@ -0,0 +1,373 @@
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	gardencore "github.com/gardener/gardener/pkg/apis/core"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/apis/seedmanagement"
+	"github.com/gardener/gardener/pkg/apis/seedmanagement/helper"
+	admissioninitializer "github.com/gardener/gardener/pkg/apiserver/admission/initializer"
+	coreclientset "github.com/gardener/gardener/pkg/client/core/clientset/internalversion"
+	clientkubernetes "github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/operation/common"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	// PluginName is the name of this admission plugin.
+	PluginName = "ManagedSeedValidator"
+)
+
+// Register registers a plugin.
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return New()
+	})
+}
+
+// ValidateManagedSeed contains listers and and admission handler.
+type ValidateManagedSeed struct {
+	*admission.Handler
+	coreClient coreclientset.Interface
+	kubeClient kubernetes.Interface
+	readyFunc  admission.ReadyFunc
+}
+
+var (
+	_ = admissioninitializer.WantsInternalCoreClientset(&ValidateManagedSeed{})
+	_ = admissioninitializer.WantsKubeClientset(&ValidateManagedSeed{})
+
+	readyFuncs = []admission.ReadyFunc{}
+)
+
+// New creates a new ValidateManagedSeed admission plugin.
+func New() (*ValidateManagedSeed, error) {
+	return &ValidateManagedSeed{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}, nil
+}
+
+// AssignReadyFunc assigns the ready function to the admission handler.
+func (v *ValidateManagedSeed) AssignReadyFunc(f admission.ReadyFunc) {
+	v.readyFunc = f
+	v.SetReadyFunc(f)
+}
+
+// SetInternalCoreClientset sets the garden core clientset.
+func (v *ValidateManagedSeed) SetInternalCoreClientset(c coreclientset.Interface) {
+	v.coreClient = c
+}
+
+// SetKubeClientset gets the clientset from the Kubernetes client.
+func (v *ValidateManagedSeed) SetKubeClientset(c kubernetes.Interface) {
+	v.kubeClient = c
+}
+
+// ValidateInitialization checks whether the plugin was correctly initialized.
+func (v *ValidateManagedSeed) ValidateInitialization() error {
+	if v.coreClient == nil {
+		return errors.New("missing garden core client")
+	}
+	if v.kubeClient == nil {
+		return errors.New("missing kubernetes client")
+	}
+	return nil
+}
+
+var _ admission.MutationInterface = &ValidateManagedSeed{}
+
+// Admit validates and if appropriate mutates the given managed seed against the shoot that it references.
+func (v *ValidateManagedSeed) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
```

Comment:
Can you add a check ensuring, that a given shoot is not registered as a ManagedSeed twice?

Reply:
Good idea, will do it.